### PR TITLE
perf(relayer): optimize ICA reveal polling for all messages

### PR DESCRIPTION
**Pull request by troykessler**
## Summary

- The 1s ICA reveal polling introduced in #8836 was accidentally gated behind `fail_fast`, so it only fired for relay API messages. Normal relayer messages fell through to `on_reprepare(ErrorEstimatingGas)` which applies the standard backoff (5s on first retry, 10s on second, etc.), causing reveal messages to arrive ~3–5s after commit on normal relayer paths.
- Removed the `self.fail_fast &&` guard so the fast polling applies to all messages.
- Replaced the single `REVEAL_POLL_INTERVAL/MAX` with a two-phase schedule:
  - **Fast phase** (attempts 1–30): 1s interval → first 30s
  - **Slow phase** (attempts 31–66): 5s interval → next 3min
  - After 66 total attempts: falls through to standard exponential backoff
- `ica_reveal_attempts` continues to track separately from `num_retries` so ICA waits don't burn the retry budget.

## Test plan

- [x] Deploy to staging and observe commit/reveal timing — reveal should land within ~1s of commit on fast chains
- [x] Verify that a reveal that genuinely can't be confirmed (e.g. commit dropped) eventually falls through to standard backoff after ~3.5min

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by troykessler** [comment]
@claude do a review please

**Comment by troykessler** [comment]
@claude, can you do a follow up review please?

**Comment by troykessler** [comment]
> **Claude finished @troykessler's task in 2m 2s** —— [View job](https://github.com/hyperlane-xyz/hyperlane-monorepo/actions/runs/27328637236)
> 
> ### Follow-up review
> * [x]  Read changed files & gather context
> * [x]  Run review (correctness + standards)
> * [x]  Post consolidated review
> 
> Since the last review, two commits landed: `2562bb28` (skip CCIP re-fetch on retries via the new cached-metadata exception) and `65532ee` (`delay_reprepare` now updates `last_attempted_at`). Both are good — the cache-exception path (`pending_message.rs:330-351`) means only the _cached_ estimate runs each poll cycle, so CCIP/metadata isn't re-fetched every 1s. The `last_attempted_at` note from the first review is resolved. 👍
> 
> Remaining items below.
> 
> #### 🟠 1. The reset-after-cap oscillation is still unaddressed (blocking-ish)
> This is the same finding from the first review (and CodeRabbit's outside-diff comment), and the latest commits don't fix it. Traced against current code:
> 
> * Cap reached → cached estimate hits the `Err(_)` arm at `:352` → `clear_metadata()`, `tx_cost_estimate = None`.
> * Metadata is rebuilt (`:366`, re-fetches CCIP) → second estimate at `:389` fails with the same `ICA: Invalid Reveal` → `attempts >= max` so the `if` at `:405` is false → **`self.ica_reveal_attempts = 0` at `:427`** → `on_reprepare`.
> * Next prepare cycle starts at `attempts = 0` → re-enters the 30×1s + 36×5s burst again, indefinitely.
> 
> So a reveal that genuinely never confirms (commit dropped — your second test-plan item) keeps emitting full ~3.5min polling bursts forever instead of decaying into standard backoff. With the old `fail_fast` path this self-terminated (message dropped at `:961`); normal messages are never dropped, so nothing terminates it.
> 
> Fix: only reset on success / non-ICA errors. Guard the `:427` reset so an ICA-invalid-reveal-at-cap leaves `ica_reveal_attempts` pinned at the max, keeping future cycles on standard backoff:
> 
> ```rust
> Err(e) => {
>     let is_invalid_reveal = is_ica_invalid_reveal(&e);
>     if is_invalid_reveal
>         && self.ica_reveal_attempts < REVEAL_POLL_FAST_MAX + REVEAL_POLL_SLOW_MAX
>     {
>         // ... existing re-queue ...
>     }
>     // Don't reset on a still-failing reveal at cap, or it re-enters the fast phase next cycle.
>     if !is_invalid_reveal {
>         self.ica_reveal_attempts = 0;
>     }
>     // ... on_reprepare ...
> }
> ```
> 
> (Success at `:396` still resets to 0, so a late-but-eventual confirmation recovers cleanly.) [Fix this →](https://claude.ai/code?q=In%20rust%2Fmain%2Fagents%2Frelayer%2Fsrc%2Fmsg%2Fpending_message.rs%20on%20branch%20feat%2Foptimize-reveal-polling%2C%20guard%20the%20ica_reveal_attempts%3D0%20reset%20at%20line%20427%20so%20it%20only%20resets%20for%20non-ICA-invalid-reveal%20errors%2C%20preventing%20the%20fast%2Fslow%20poll%20cycle%20from%20restarting%20after%20the%20cap%20is%20reached&repo=hyperlane-xyz/hyperlane-monorepo)
> 
> #### 🟡 2. The fast/slow re-queue block is now duplicated verbatim
> The increment + interval/phase select + `warn!` + `delay_reprepare` block is identical in two places (`:335-350` and `:409-425`). Now that it has grown, extract a helper to keep the two paths in sync:
> 
> ```rust
> fn requeue_ica_reveal(&mut self) -> PendingOperationResult {
>     self.ica_reveal_attempts += 1;
>     let (interval, phase) = if self.ica_reveal_attempts <= REVEAL_POLL_FAST_MAX {
>         (REVEAL_POLL_FAST_INTERVAL, "fast")
>     } else {
>         (REVEAL_POLL_SLOW_INTERVAL, "slow")
>     };
>     warn!(/* … phase, interval_secs, max_attempts … */);
>     self.delay_reprepare(interval, ReprepareReason::AwaitingIcaReveal)
> }
> ```
> 
> Both call sites collapse to `if is_ica_invalid_reveal(&e) && self.ica_reveal_attempts < REVEAL_POLL_TOTAL_MAX { return self.requeue_ica_reveal(); }`.
> 
> #### 🟡 3. `mainnet3/agent.ts` CORS preview URL — still unresolved
> `agent.ts:874` still adds the Vercel **preview** URL (`hyperlane-warp-template-git-xaroz-swap-form-abacus-works.vercel.app`) to the mainnet3 relay-API `corsOrigins`. Same question as last time: is this meant to ship permanently to prod? A `git-<branch>` preview URL disappears when the branch is deleted, and it's unrelated to the reveal-polling change. If it's for testing, it probably shouldn't land in mainnet3; if permanent, a stable domain is safer.
> 
> #### 🟢 4. Minor (cosmetic, unchanged from first review)
> * `REVEAL_POLL_FAST_MAX + REVEAL_POLL_SLOW_MAX` is now spelled out **4×**. A `const REVEAL_POLL_TOTAL_MAX: u32 = REVEAL_POLL_FAST_MAX + REVEAL_POLL_SLOW_MAX;` would DRY it and make the "66 total" intent explicit.
> * `REVEAL_POLL_SLOW_MAX = 36` is a phase _count_, not a ceiling like `FAST_MAX` reads. `*_SLOW_COUNT` (or the total const above) reduces ambiguity.
> * Still no unit test for the fast→slow boundary (attempt 30 vs 31), where the `<=`/`<` logic is off-by-one sensitive. Cheap insurance given this is the core of the PR.
> 
> Arithmetic checks out (fast 1–30 @1s = 30s, slow 31–66 @5S = 180s; cap at `< 66`). #1 is the one I'd want fixed before merge — the others are cleanup / a config decision.

the new cors preview url is intentional and should stay, fixed the other issues, take a final look @claude

File: rust/main/agents/relayer/src/msg/pending_message.rs [hunk 1]
```diff
@@ -322,6 +328,12 @@ impl PendingOperation for PendingMessage {
                         tracing::debug!(USE_CACHE_METADATA_LOG);
                         Some(s)
```
Comment (paulbalaji) [inline rust/main/agents/relayer/src/msg/pending_message.rs:330]: Minor (non-blocking): the cached-metadata success path returns `Some(s)` without resetting `ica_reveal_attempts`, unlike the rebuilt-metadata success arm which resets to `0`. Once a reveal confirms via cached metadata, the counter stays pinned. If the same message later re-fails with `ICA: Invalid Reveal` (e.g. the commit gets reorged out after confirmation), polling resumes mid-window — or skips it entirely if pinned at the cap — instead of restarting the fast phase.

For symmetry with the reset on the rebuilt path, consider:
```rust
Ok(s) => {
    self.ica_reveal_attempts = 0;
    tracing::debug!(USE_CACHE_METADATA_LOG);
    Some(s)
}
```
Edge case, so take it or leave it.

**Review by paulbalaji — APPROVED** [review]
## Review — `perf(relayer): optimize ICA reveal polling`

Reviewed at `fe2928a1`. Independent pass (Rust correctness, infra/CORS config, build/clippy/tests), accounting for the prior Claude and CodeRabbit reviews.

**Verified resolved / correct:**
- **Reset-after-cap oscillation** (raised by both prior reviews): genuinely fixed. Traced the cap path — at `REVEAL_POLL_TOTAL_MAX` with a still-failing reveal, the `if !is_ica_invalid_reveal(&e)` guard keeps `ica_reveal_attempts` pinned at 66, so subsequent prepares go straight to standard backoff instead of re-entering the fast burst. No remaining path re-zeros the counter while the reveal keeps failing.
- **Two-phase arithmetic**: correct, no off-by-one. Attempts 1–30 @1s (30s) → 31–66 @5s (180s); matches the comments.
- **`delay_reprepare` setting `last_attempted_at`**: safe — only consumed by a `Display` log field; queue ordering reads `next_attempt_after`. Fixes a previously-stale log value.
- **CI clippy**: green. The helper uses `saturating_add`, satisfying the relayer's `#![deny(clippy::arithmetic_side_effects)]`. (An earlier draft of this review flagged a `+= 1` clippy failure — that was on an older commit and is now resolved.)
- **`corsOrigins` comma-list**: parses correctly (rust splits on `,`, validates each origin individually). Functional.

**Non-blocking notes:**
- One inline comment below — the cached-metadata success path doesn't reset `ica_reveal_attempts` (asymmetric with the rebuilt-path reset). Low-probability edge case, trivial fix; flagging for completeness.
- **No tests** cover the new reveal-polling logic (fast→slow boundary at attempt 30/31, the cap, and the reset-at-cap pinning). The `<=`/`<` boundary is off-by-one-sensitive — a small `requeue_ica_reveal`-driven boundary test would be cheap insurance, but not required.
- **CORS preview URL** (`agent.ts:874`): the Vercel branch-preview origin in mainnet3 prod is acknowledged as intentional by the author. Blast radius is small (the `/relay` endpoint only re-submits already-on-chain messages, globally rate-limited; CORS is browser-only). Worth tracking for removal once the `xaroz/swap-form` branch is gone, since `git-<branch>` preview URLs are ephemeral.
- **Changeset**: none present; `@hyperlane-xyz/infra` is private and CI doesn't gate `version:check`, so not required here.

LGTM — approving. Core relayer change is sound and all previously-flagged issues are resolved.
